packages/salesforcedx-vscode-core/src/conflict/conflictDetectionService.ts

```diff
@@ -0,0 +1,248 @@
+import * as vscode from 'vscode';
+import { channelService } from '../channels';
+import { nls } from '../messages';
+import { notificationService, ProgressNotification } from '../notifications';
+import { taskViewService } from '../statuses';
+import { telemetryService } from '../telemetry';
+import { getRootWorkspacePath } from '../util';
+import {
+  CommonDirDirectoryDiffer,
+  DirectoryDiffer,
+  DirectoryDiffResults
+} from './directoryDiffer';
+
+export interface ConflictDetectionConfig {
+  usernameOrAlias: string;
+  packageDir: string;
+  manifest: string;
+}
+
+export class ConflictDetector {
+  private differ: DirectoryDiffer;
+  private static instance: ConflictDetector;
+
+  constructor(differ?: DirectoryDiffer) {
+    this.differ = differ || new CommonDirDirectoryDiffer();
+  }
+
+  public static getInstance(): ConflictDetector {
+    if (!ConflictDetector.instance) {
+      ConflictDetector.instance = new ConflictDetector(
+        new CommonDirDirectoryDiffer()
+      );
+    }
+    return ConflictDetector.instance;
+  }
+
+  private readonly relativeMetdataTempPath = path.join(
+    '.sfdx',
+    'tools',
+    'conflicts'
+  );
+
+  private readonly relativePackageXmlPath = path.join(
+    this.relativeMetdataTempPath,
+    'package.xml'
+  );
+
+  private readonly relativeUnpackagedPath = path.join(
+    this.relativeMetdataTempPath,
+    'unpackaged'
+  );
+
+  private readonly relativeConvertedPath = path.join(
+    '.sfdx',
+    'tools',
+    'conflicts',
+    'converted'
+  );
+
+  public buildRetrieveOrgSourceCommand(data: ConflictDetectionConfig): Command {
+    return new SfdxCommandBuilder()
+      .withDescription(nls.localize('conflict_detect_retrieve_org_source'))
+      .withArg('force:mdapi:retrieve')
+      .withFlag('--retrievetargetdir', this.relativeMetdataTempPath)
+      .withFlag('--unpackaged', this.relativePackageXmlPath)
+      .withFlag('--targetusername', data.usernameOrAlias)
+      .withLogName('conflict_detect_retrieve_org_source')
+      .build();
+  }
+
+  public buildMetadataApiConvertOrgSourceCommand(
+    data: ConflictDetectionConfig
+  ): Command {
+    return new SfdxCommandBuilder()
+      .withDescription(nls.localize('conflict_detect_convert_org_source'))
+      .withArg('force:mdapi:convert')
+      .withFlag('--rootdir', this.relativeUnpackagedPath)
+      .withFlag('--outputdir', this.relativeConvertedPath)
+      .withLogName('conflict_detect_convert_org_source')
+      .build();
+  }
+
+  public async checkForConflicts(
+    data: ConflictDetectionConfig
+  ): Promise<DirectoryDiffResults> {
+    const cancellationTokenSource = new vscode.CancellationTokenSource();
+    const cancellationToken = cancellationTokenSource.token;
+
+    const results = await this.checkForConflictsInternal(
+      getRootWorkspacePath(),
+      data,
+      cancellationTokenSource,
+      cancellationToken
+    );
+    return results;
+  }
+
+  private async checkForConflictsInternal(
```

Comment:
The logic here could be broke down into different methods like one that just creates a directory, another that unzips the metadata zip file, etc. This will allow you to write unit tests per method and have a simpler integration test. Similar to how we've done it in other areas like the org browser.

Reply:
I tried breaking it out, but I ended up putting it back the way it was. I separated out the error handling to make it a little cleaner. Since I can intercept the calls to executeCommand the integration test isn't too bad. I feel better covering this with an integration test anyway. I'll break out the directory diff-ing into a separate test suite, too.